<a href="https://colab.research.google.com/github/SupawadeeKhamthuy/DWDM21/blob/main/MiniExam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### ข้อสอบ

จาก data ```flights.csv``` และ ```airports.csv``` ใน https://drive.google.com/drive/folders/1EFo1_uQgWkRiXjezL0T4YwteYrsqU5MV?usp=sharing
0. หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด
1. หารัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดและน้อยที่สุด
2. วาด Histogram แสดงจำนวนเครื่องบินขาออกของแต่ละวันในสัปดาห์
3. สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด
4. สายการบินอะไรมีความล่าช้ามากที่สุดและน้อยที่สุด
5. ช่วงเช้าหรือช่วงบ่ายมีเครื่องบินขาเข้ามากกว่ากัน

ให้นักศึกษาทำข้อสอบ 2 ข้อจากที่แสดงข้างบนโดยกรอกรหัสนักศึกษาลงใน function ```gen_exams()``` แล้ว run เพื่อให้ได้ข้อสอบที่นักศึกษาต้องทำ 2 ข้อ (ข้อละ 3 คะแนน)
และให้นักศึกษาคิดโจทย์เองอีก 1 ข้อ (4 คะแนน)


In [1]:
import pandas as pd

In [2]:
from google.colab import drive 
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os

In [4]:
flights = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/Mini Exam/flights.csv')
flights

วันในเดือน  วันในสัปดาห์  ... ความล่าช้าขาออก  ความล่าช้าขาเข้า
0                19             5  ...              -3                 1
1                19             5  ...               0                -8
2                19             5  ...              -4               -15
3                19             5  ...              28                24
4                19             5  ...              -6               -11
...             ...           ...  ...             ...               ...
2702213          19             6  ...               1                -3
2702214          19             6  ...              -2                -3
2702215          19             6  ...              18                 9
2702216          19             6  ...              10                 7
2702217          19             6  ...               0               -13

[2702218 rows x 7 columns]

In [5]:
airports = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/Mini Exam/airports.csv')
airports

รหัสสนามบิน         เมือง รัฐ                          ชื่อสนามบิน
0          10165   Adak Island  AK                                 Adak
1          10299     Anchorage  AK  Ted Stevens Anchorage International
2          10304         Aniak  AK                        Aniak Airport
3          10754        Barrow  AK      Wiley Post/Will Rogers Memorial
4          10551        Bethel  AK                       Bethel Airport
..           ...           ...  ..                                  ...
360        11233      Cheyenne  WY  Cheyenne Regional/Jerry Olson Field
361        11097          Cody  WY                 Yellowstone Regional
362        11865      Gillette  WY             Gillette Campbell County
363        12441       Jackson  WY                         Jackson Hole
364        14543  Rock Springs  WY       Rock Springs Sweetwater County

[365 rows x 4 columns]

In [7]:
def gen_exams(ID):
    import math
    a = int(ID[-1])%6
    b = int(ID[-1])%3
    print(f'รหัสนักศึกษา {ID} ทำข้อ {a} และ {b if a != b else 5-b}')

ตัวอย่างการ run

In [ ]:
gen_exams('62123456-9')

รหัสนักศึกษา 62123456-9 ทำข้อ 3 และ 0


In [14]:
gen_exams('623020544-6')

รหัสนักศึกษา 623020544-6 ทำข้อ 0 และ 5


In [20]:
merged_table = flights.merge(airports,how='left',left_on='รหัสสนามบินต้นทาง',right_on='รหัสสนามบิน') ## รวม 2 ตารางเข้าด้วยกัน
merged_table

วันในเดือน  ...                                 ชื่อสนามบิน
0                19  ...                  Detroit Metro Wayne County
1                19  ...                Salt Lake City International
2                19  ...                      Portland International
3                19  ...             Lambert-St. Louis International
4                19  ...  Cincinnati/Northern Kentucky International
...             ...  ...                                         ...
2702213          19  ...                       Orlando International
2702214          19  ...    Hartsfield-Jackson Atlanta International
2702215          19  ...           Albuquerque International Sunport
2702216          19  ...    Hartsfield-Jackson Atlanta International
2702217          19  ...    Hartsfield-Jackson Atlanta International

[2702218 rows x 11 columns]

In [21]:
merged_table.isnull().any()

วันในเดือน            False
วันในสัปดาห์          False
สายการบิน             False
รหัสสนามบินต้นทาง     False
รหัสสนามบินปลายทาง    False
ความล่าช้าขาออก       False
ความล่าช้าขาเข้า      False
รหัสสนามบิน           False
เมือง                 False
รัฐ                   False
ชื่อสนามบิน           False
dtype: bool

0.  หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด

In [23]:
merged_table.groupby('รัฐ').count()

วันในเดือน  วันในสัปดาห์  สายการบิน  ...  รหัสสนามบิน   เมือง  ชื่อสนามบิน
รัฐ                                       ...                                  
AK         7700          7700       7700  ...         7700    7700         7700
AZ       100710        100710     100710  ...       100710  100710       100710
CA       380046        380046     380046  ...       380046  380046       380046
CO        97259         97259      97259  ...        97259   97259        97259
CT        13042         13042      13042  ...        13042   13042        13042
DC        80536         80536      80536  ...        80536   80536        80536
FL       208139        208139     208139  ...       208139  208139       208139
GA       148563        148563     148563  ...       148563  148563       148563
HI        28395         28395      28395  ...        28395   28395        28395
IL       176939        176939     176939  ...       176939  176939       176939
IN        18038         18038      18038  ...        18038   18038        18038
KY        10915         10915      10915  ...        10915   10915        10915
LA        25184         25184      25184  ...        25184   25184        25184
MA        66699         66699      66699  ...        66699   66699        66699
MD        51761         51761      51761  ...        51761   51761        51761
MI        62584         62584      62584  ...        62584   62584        62584
MN        60174         60174      60174  ...        60174   60174        60174
MO        63570         63570      63570  ...        63570   63570        63570
NC       104766        104766     104766  ...       104766  104766       104766
NE        13487         13487      13487  ...        13487   13487        13487
NJ        64313         64313      64313  ...        64313   64313        64313
NM        17542         17542      17542  ...        17542   17542        17542
NV        88222         88222      88222  ...        88222   88222        88222
NY       132823        132823     132823  ...       132823  132823       132823
OH        66389         66389      66389  ...        66389   66389        66389
OK        13922         13922      13922  ...        13922   13922        13922
OR        30484         30484      30484  ...        30484   30484        30484
PA        67957         67957      67957  ...        67957   67957        67957
PR        13840         13840      13840  ...        13840   13840        13840
RI         9313          9313       9313  ...         9313    9313         9313
TN        53414         53414      53414  ...        53414   53414        53414
TX       275503        275503     275503  ...       275503  275503       275503
UT        47324         47324      47324  ...        47324   47324        47324
VA        21491         21491      21491  ...        21491   21491        21491
WA        60172         60172      60172  ...        60172   60172        60172
WI        21002         21002      21002  ...        21002   21002        21002

[36 rows x 10 columns]

In [24]:
Ans = merged_table.groupby('รัฐ').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False) # เลือก sort_values by column ที่ไม่มี missing มาใช้
Ans

วันในเดือน  วันในสัปดาห์  สายการบิน  ...  รหัสสนามบิน   เมือง  ชื่อสนามบิน
รัฐ                                       ...                                  
CA       380046        380046     380046  ...       380046  380046       380046
TX       275503        275503     275503  ...       275503  275503       275503
FL       208139        208139     208139  ...       208139  208139       208139
IL       176939        176939     176939  ...       176939  176939       176939
GA       148563        148563     148563  ...       148563  148563       148563
NY       132823        132823     132823  ...       132823  132823       132823
NC       104766        104766     104766  ...       104766  104766       104766
AZ       100710        100710     100710  ...       100710  100710       100710
CO        97259         97259      97259  ...        97259   97259        97259
NV        88222         88222      88222  ...        88222   88222        88222
DC        80536         80536      80536  ...        80536   80536        80536
PA        67957         67957      67957  ...        67957   67957        67957
MA        66699         66699      66699  ...        66699   66699        66699
OH        66389         66389      66389  ...        66389   66389        66389
NJ        64313         64313      64313  ...        64313   64313        64313
MO        63570         63570      63570  ...        63570   63570        63570
MI        62584         62584      62584  ...        62584   62584        62584
MN        60174         60174      60174  ...        60174   60174        60174
WA        60172         60172      60172  ...        60172   60172        60172
TN        53414         53414      53414  ...        53414   53414        53414
MD        51761         51761      51761  ...        51761   51761        51761
UT        47324         47324      47324  ...        47324   47324        47324
OR        30484         30484      30484  ...        30484   30484        30484
HI        28395         28395      28395  ...        28395   28395        28395
LA        25184         25184      25184  ...        25184   25184        25184
VA        21491         21491      21491  ...        21491   21491        21491
WI        21002         21002      21002  ...        21002   21002        21002
IN        18038         18038      18038  ...        18038   18038        18038
NM        17542         17542      17542  ...        17542   17542        17542
OK        13922         13922      13922  ...        13922   13922        13922
PR        13840         13840      13840  ...        13840   13840        13840
NE        13487         13487      13487  ...        13487   13487        13487
CT        13042         13042      13042  ...        13042   13042        13042
KY        10915         10915      10915  ...        10915   10915        10915
RI         9313          9313       9313  ...         9313    9313         9313
AK         7700          7700       7700  ...         7700    7700         7700

[36 rows x 10 columns]

รัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด

In [25]:
max = Ans.iloc[:1,[3]]
max

รหัสสนามบินต้นทาง
รัฐ                   
CA              380046

In [26]:
min = Ans.iloc[-1:,[3]]
min

รหัสสนามบินต้นทาง
รัฐ                   
AK                7700

5.  วันธรรมดาหรือวันเสาร์-อาทิตย์มากกว่ากัน  (วันจันทร์ = 1, วันอังคาร = 2, ...)

In [39]:
monday = merged_table['วันในสัปดาห์'] == 1
tuesday = merged_table['วันในสัปดาห์'] == 2
wednesday =  merged_table['วันในสัปดาห์'] == 3
thursday =  merged_table['วันในสัปดาห์']== 4
friday = merged_table['วันในสัปดาห์']== 5
normal = merged_table[monday | tuesday | wednesday | thursday | friday]
normal

วันในเดือน  ...                                 ชื่อสนามบิน
0                19  ...                  Detroit Metro Wayne County
1                19  ...                Salt Lake City International
2                19  ...                      Portland International
3                19  ...             Lambert-St. Louis International
4                19  ...  Cincinnati/Northern Kentucky International
...             ...  ...                                         ...
2702141          18  ...                Salt Lake City International
2702142          18  ...                  Detroit Metro Wayne County
2702143          18  ...                  Detroit Metro Wayne County
2702144          18  ...                Salt Lake City International
2702145          18  ...                     San Diego International

[2011453 rows x 11 columns]

In [40]:
normal.groupby('วันในสัปดาห์').count()

วันในเดือน  สายการบิน  ...     รัฐ  ชื่อสนามบิน
วันในสัปดาห์                         ...                     
1                 407837     407837  ...  407837       407837
2                 397594     397594  ...  397594       397594
3                 403072     403072  ...  403072       403072
4                 406563     406563  ...  406563       406563
5                 396387     396387  ...  396387       396387

[5 rows x 10 columns]

In [41]:
Ans_nm = normal.groupby('วันในสัปดาห์').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False) # เลือก sort_values by column ที่ไม่มี missing มาใช้
Ans_nm

วันในเดือน  สายการบิน  ...     รัฐ  ชื่อสนามบิน
วันในสัปดาห์                         ...                     
1                 407837     407837  ...  407837       407837
4                 406563     406563  ...  406563       406563
3                 403072     403072  ...  403072       403072
2                 397594     397594  ...  397594       397594
5                 396387     396387  ...  396387       396387

[5 rows x 10 columns]

In [38]:
sum_nm = Ans_nm[['รหัสสนามบินปลายทาง']].sum()
sum_nm

รหัสสนามบินปลายทาง    2011453
dtype: int64

In [43]:
saturday = merged_table['วันในสัปดาห์'] == 6
sunday = merged_table['วันในสัปดาห์'] == 7
holiday = merged_table[saturday | sunday]
holiday

วันในเดือน  วันในสัปดาห์  ... รัฐ                               ชื่อสนามบิน
31               20             6  ...  CO                      Denver International
32               20             6  ...  UT              Salt Lake City International
33               20             6  ...  NY                                 LaGuardia
34               20             6  ...  UT              Salt Lake City International
35               20             6  ...  CA                 Los Angeles International
...             ...           ...  ...  ..                                       ...
2702213          19             6  ...  FL                     Orlando International
2702214          19             6  ...  GA  Hartsfield-Jackson Atlanta International
2702215          19             6  ...  NM         Albuquerque International Sunport
2702216          19             6  ...  GA  Hartsfield-Jackson Atlanta International
2702217          19             6  ...  GA  Hartsfield-Jackson Atlanta International

[690765 rows x 11 columns]

In [44]:
holiday.groupby('วันในสัปดาห์').count()

วันในเดือน  สายการบิน  ...     รัฐ  ชื่อสนามบิน
วันในสัปดาห์                         ...                     
6                 318537     318537  ...  318537       318537
7                 372228     372228  ...  372228       372228

[2 rows x 10 columns]

In [45]:
Ans_hd = holiday.groupby('วันในสัปดาห์').count().sort_values(by=['รหัสสนามบินปลายทาง'],ascending=False) # เลือก sort_values by column ที่ไม่มี missing มาใช้
Ans_hd

วันในเดือน  สายการบิน  ...     รัฐ  ชื่อสนามบิน
วันในสัปดาห์                         ...                     
7                 372228     372228  ...  372228       372228
6                 318537     318537  ...  318537       318537

[2 rows x 10 columns]

In [46]:
sum_hd = Ans_hd[['รหัสสนามบินปลายทาง']].sum()
sum_hd

รหัสสนามบินปลายทาง    690765
dtype: int64

ความแตกต่างของวันธรรมดากับวันในสัปดาห์

In [47]:
sum_nm - sum_hd

รหัสสนามบินปลายทาง    1320688
dtype: int64

จากโจทย์ วันธรรมดา หรือวันเสาร์-อาทิตย์มีเครื่องบินขาเข้ามากกว่ากัน

วันธรรมดา มีจำนวนเครื่องบินขาเข้า เท่ากับ 2,011,453 ลำ

วันเสาร์-อาทิตย์ มีจำนวนเครื่องบินขาเข้า เท่ากับ 690,765 ลำ

ดังนั้น วันธรรมดาจะมีจำนวนเครื่องบินขาเข้ามากกว่าวันเสาร์-อาทิตย์ เท่ากับ 1,320,688 ลำ